PHPCompatibility/Sniffs/Constants/NewConstantsSniff.php

```diff
@@ -8085,6 +8089,31 @@ protected function handleFeature(File $phpcsFile, $stackPtr, array $itemInfo)
         $itemArray   = $this->newConstants[$itemInfo['name']];
         $versionInfo = $this->getVersionInfo($itemArray);
 
+        if ($itemInfo['name'] === 'T_BAD_CHARACTER') {
```

Comment:
It seems to me that this is much more logic than is needed, partially due to the functionality of the trait being broken with the change in the version array.

I think this can be greatly simplified by undoing the change to the version array (and possibly adding a `// Note: this constant has special casing in the handleFeature() method as it was also present in PHP < 7.0.` comment instead) and moving this block to below the `empty($versionInfo['not_in_version']) || ScannedCode::shouldRunOnOrBelow($versionInfo['not_in_version']) === false` condition block.

Proposed simplification:
```php
        // T_BAD_CHARACTER is a special case. It was removed in 7.0.0 and re-added in 7.4.0
        // See also PHPCompatibility.Constants.RemovedConstants
        if ($itemInfo['name'] === 'T_BAD_CHARACTER') {
            if (ScannedCode::shouldRunOnOrAbove('7.0') === true) {
                $message = 'The constant "T_BAD_CHARACTER" is not present in PHP versions 7.0 through 7.3';
                $msgInfo = $this->getMessageInfo($itemInfo['name'], $itemInfo['name'], $versionInfo);

                $phpcsFile->addError($message, $stackPtr, $msgInfo['errorcode'], $msgInfo['data']);
            }

            return;
        }
```

This will work as the `empty($versionInfo['not_in_version']) || ScannedCode::shouldRunOnOrBelow($versionInfo['not_in_version']) === false` condition block already filters out any the "code base only supports PHP 7.4 or higher" case.

So the logic I propose above, just needs to verify if PHP >= 7.0 needs to be supported. If so, throw an error. If not, stay silent as in that case, it means that the code base only supports PHP <= 5.6.

Reply:
Thanks for the suggestion. Upon reflection, this makes sense. I've applied this change and the test suite still passes.